apps/opt.c

```diff
@@ -694,32 +692,66 @@ int opt_next(void)
             return -1;
         case '>':
             /* Output file. */
-#if !defined(_WIN32)
-            c = OPENSSL_strdup(arg);
-            if (c == NULL) {
-                BIO_printf(bio_err,
-                           "%s: Memory allocation failure\n", prog);
-                return -1;
-            }
-            oerrno = errno;
-            errno = 0;
-            if (strcmp(arg, "-") == 0
-                || (app_access(app_dirname(c), W_OK) == 0
-                    && app_isdir(arg) <= 0
-                    && (app_access(arg, W_OK) == 0 || errno == ENOENT))) {
-                OPENSSL_free(c);
+
+            /*
+             * "-" indicates we output to stdout, nothing more needs to be done
+             */
+            if (strcmp(arg, "-") == 0)
                 break;
+#ifdef __VMS
+            {
+                struct stat st;
+
+                /*
+                 * On VMS, "character devices" (such as NLA0:) aren't possible
+                 * to get any sensible information on with access(), because
+                 * the underlying system libraries and services do not support
+                 * the necessary operation to determine access on devices.
+                 * We therefore simply check for these "character devices" and
+                 * simply assumes they are writable here, and the attempt to
+                 * actually write to them will have to succeed or fail later.
+                 */
+                if (stat(arg, &st) == 0 && S_ISCHR(st.st_mode))
+                    break;
             }
-            OPENSSL_free(c);
-            if (errno == 0)
-                /* only possible if 'arg' is a directory */
-                estr = "is a directory";
-            else
+#endif
+#if !defined(_WIN32)
+            /*
+             * If the intended output file exists, we start with checking if it
+             * exists.
+             */
+            if (app_access(arg, F_OK) == 0) {
+                /*
+                 * If it exists, we need to make sure it isn't a directory and
+                 * that it's writable.
+                 */
+                if (app_isdir(arg) > 0)
+                    estr = "is a directory";
+                else if (app_access(arg, W_OK) == 0)
+                    break;
+                else
+                    estr = strerror(errno);
+            } else {
+                /*
+                 * If the intended output file doesn't exist, we check if the
+                 * directory it's going to be written in is writable (which
+                 * implies it exists).
+                 */
+                char *c = OPENSSL_strdup(arg);
+                if (c == NULL) {
+                    BIO_printf(bio_err,
+                               "%s: Memory allocation failure\n", prog);
+                    return -1;
+                }
+                if (app_access(app_dirname(c), W_OK) == 0) {
```

Comment:
Does it need a check for the `arg` not to be a directory also here, when it's going to create a new file?

So the logic would be:

1. if the `-out`'s value exists, check: a) not be a directory and b) itself is writable;
2. if the `-out`'s value doesn't exist, check: a) not be a directory and b) its parent is writable

And the "a" section of both conditions could be extracted out...

Reply:
I don't see why...  in this part of the code, `access(arg, F_OK)` has determined that `arg` isn't an existing file at all, so it cannot be a directory.  But maybe you mean that we should check that the parent directory actually is a directory?

Reply:
Nah, I meant this scenario: take `arg` as something like `/usr/non-exist-name/`, it doesn't exist. So with current logic, it will be checked only for if its parent dir (the `/usr`) is writable and considers the `arg` as valid. But `/usr/non-exist-name/` is not a valid filename to use with `-out` option.

Hmmm, and for the check of parent directory, I didn't consider this then, but seems `app_access(app_dirname(c), W_OK)` has covered the check for existence...right?

Reply:
> Hmmm, and for the check of parent directory, I didn't consider this then, but seems `app_access(app_dirname(c), W_OK)` has covered the check for existence...right?

Quite right

Reply:
Regarding things like `-out random4711foo/`, you're right, that doesn't get caught here...  but then I wonder how far this should be taken.

Reply:
I think it should keep consistent with the original check logic - which always checks if the `-out` value inputed directs to a directory or not...

Reply:
Huh?  Please tell me how `stat` (which `app_isdir` uses) can determine the mode of a file that doesn't exist?

Reply:
In your original check, the `app_isdir(arg)` clearly *assumes* that `arg` contains the path of an existing file.

Reply:
Ahh! I know the problem, I tested this check with `genrsa` which calls the `bio_open_owner` earlier than doing the true-job, and then that `bio_open_owner` launched an error...I was misled and sorry for the confusion...So you are right, the original logic lacks of checking for creating a directory...

But it seems it's very easy to fix by adding a check for the trailing '/' on Unix-like platform:

```
+#if !defined(_WIN32)
+            /*
+             * If the intended output file exists, we start with checking if it
+             * exists.
+             */
+            if (*(arg + strlen(arg) - 1) == '/') {
+                estr = "is a directory";
+            } else if (app_access(arg, F_OK) == 0) {
```

What do you think about?

 (BTW, I think all `bio_open_owner` stuff should be moved after the 'true-jobs' in apps like `genrsa`, I would like to do so by opening a new PR)

Reply:
I'm not sure I understand...  if `arg` is `"/foo/bar"` and that's a directory, that check will fail.